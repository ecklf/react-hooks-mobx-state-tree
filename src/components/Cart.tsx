import React, { useState, ChangeEvent } from "react";
import { observer } from "mobx-react-lite";

import { useMst } from "../models/Root";

import Button from "./Button";

interface Props {}

const Cart: React.FC<Props> = observer(() => {
  const { cart } = useMst();

  const [name, setName] = useState("The Hobbit");
  const [price, setPrice] = useState(20);

  const isDisabled = () => {
    return name === "" || price === 0;
  };

  return (
    <div className="w-64 mx-auto mt-16">
      <p className="font-bold text-2xl text-center">Item Cart</p>
      <label className="block">
        <span className="text-gray-200">Name</span>
        <input
          className="form-input mt-1 block w-full"
          value={name}
          onChange={(e: ChangeEvent<HTMLInputElement>) => {
            if (e && e.target) {
              setName(e.target.value);
            }
          }}
        />
      </label>
      <label className="block mt-2">
        <span className="text-gray-200">Price</span>
        <input
          className="form-input mt-1 block w-full"
          type="number"
          min="0.0"
          step="any"
          value={price}
          onChange={(e: ChangeEvent<HTMLInputElement>) => {
            if (e && e.target) {
              setPrice(Number(e.target.value));
            }
          }}
        />
      </label>
      <div className="mt-4 flex items-center">
        <p
          style={{ fontVariant: "tabular-nums" }}
          className="leading-tight font-bold text-lg"
        >
          Items: {cart.totalItems}
        </p>
        <p
          style={{ fontVariant: "tabular-nums" }}
          className="ml-4 flex-grow leading-tight font-bold text-lg"
        >
          Total: {cart.totalPrice}
        </p>
        <Button
          disabled={isDisabled()}
          label="Add"
          className={`ml-4 ${isDisabled() ? "cursor-not-allowed" : ""}`}
          onClick={() => {
            cart.addCartItem({
              name,
              price
            });
          }}
        />
      </div>
      <div className="my-12 h-64 bg-gray-800 rounded overflow-y-scroll">
        {cart.items.map((item, index) => {
          return (
            <div
              key={`${item.name}-${index}`}
              className={`${index !== 0 &&
                "mt-2"} mr-2 px-4 py-2 flex items-center rounded rounded-sm leading-tight text-gray-700 bg-gray-200`}
            >
              <div className="flex-grow">{item.name}</div>
              <div>{item.price}</div>
              <span
                className="ml-5 cursor-pointer select-none"
                onClick={item.remove}
                role="img"
                aria-label="delete"
              >
                ‚ùå
              </span>
            </div>
          );
        })}
      </div>
    </div>
  );
});

export default Cart;
