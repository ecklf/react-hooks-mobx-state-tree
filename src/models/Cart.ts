import {
  types,
  Instance,
  SnapshotIn,
  getParent,
  destroy
} from "mobx-state-tree";

export const Tag = types.model({
  name: types.string
});

export const CartItem = types
  .model({
    name: types.string,
    price: types.number,
    tags: types.optional(types.array(Tag), [])
  })
  .actions(self => ({
    changeName(newName: string) {
      self.name = newName;
    },
    changePrice(newPrice: number) {
      self.price = newPrice;
    },
    remove() {
      getParent<typeof Cart>(self, 2).remove(self);
    }
  }));

export const Cart = types
  .model({
    items: types.optional(types.array(CartItem), [])
  })
  .actions(self => ({
    addCartItem(
      cartItem: SnapshotIn<typeof CartItem> | Instance<typeof CartItem>
    ) {
      self.items.push(cartItem);
    },
    remove(item: SnapshotIn<typeof CartItem>) {
      destroy(item);
    }
  }))
  .views(self => ({
    get totalItems() {
      return self.items.length;
    },
    get totalPrice() {
      return self.items.reduce((sum, entry) => sum + entry.price, 0);
    }
  }));
