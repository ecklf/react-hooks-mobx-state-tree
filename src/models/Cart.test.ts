import { CartItem, Cart } from "./Cart";
import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { reaction } from "mobx";

it("should create a model instance", () => {
  const item = CartItem.create({
    name: "Google Pixel 3",
    price: 499
  });
  const patches: any = [];

  onPatch(item, snapshot => {
    patches.push(snapshot);
  });

  item.changeName("Google Pixel 4");
  item.changePrice(799);
  expect(patches).toMatchSnapshot();
});

it("should create a cart instance", () => {
  const cart = Cart.create({
    items: [
      {
        name: "Instance Item",
        price: 2.83,
        tags: [{ name: "first" }]
      }
    ]
  });
  expect(getSnapshot(cart)).toMatchSnapshot();
});

it("can add new items to the cart", () => {
  const cart = Cart.create();
  const states: any = [];

  onSnapshot(cart, snapshot => {
    states.push(snapshot);
  });

  cart.addCartItem({
    name: "New Item 1",
    price: 1
  });
  cart.addCartItem({
    name: "New Item 2",
    price: 2
  });
  expect(states).toMatchSnapshot();
});

it("can remove an item from the cart", () => {
  const cart = Cart.create();
  const states: any = [];

  onSnapshot(cart, snapshot => {
    states.push(snapshot);
  });

  cart.addCartItem({
    name: "New Item 1",
    price: 1
  });
  cart.addCartItem({
    name: "New Item 2",
    price: 2
  });
  cart.items[0].remove();
  cart.items[0].remove();
  expect(states).toMatchSnapshot();
});

it("can rename an item in the cart", () => {
  const cart = Cart.create();
  const patches: any = [];

  onPatch(cart, snapshot => {
    patches.push(snapshot);
  });

  cart.addCartItem({
    name: "New Item 1",
    price: 1
  });

  cart.items[0].changeName("New Item (new)");
  expect(patches).toMatchSnapshot();
});

it("can return the total items of the cart", () => {
  const cart = Cart.create();

  cart.addCartItem({
    name: "Item 1",
    price: 1
  });

  expect(cart.totalItems).toBe(1);
});

it("can calculate the total price of the cart", () => {
  const cart = Cart.create();

  cart.addCartItem({
    name: "Item 1",
    price: 1
  });
  cart.addCartItem({
    name: "Item 2",
    price: 1
  });

  expect(cart.totalPrice).toBe(2);
  let changed = 0;
  reaction(() => cart.totalPrice, () => changed++);

  expect(changed).toBe(0);
  cart.items[0].changeName("Item 1 (new)");
  expect(changed).toBe(0);

  cart.items[0].changePrice(2);
  expect(changed).toBe(1);
  expect(cart.totalPrice).toBe(3);
});
