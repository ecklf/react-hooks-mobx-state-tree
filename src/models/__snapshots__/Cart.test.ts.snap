// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items to the cart 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "name": "New Item 1",
        "price": 1,
        "tags": Array [],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "name": "New Item 1",
        "price": 1,
        "tags": Array [],
      },
      Object {
        "name": "New Item 2",
        "price": 2,
        "tags": Array [],
      },
    ],
  },
]
`;

exports[`can remove an item from the cart 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "name": "New Item 1",
        "price": 1,
        "tags": Array [],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "name": "New Item 1",
        "price": 1,
        "tags": Array [],
      },
      Object {
        "name": "New Item 2",
        "price": 2,
        "tags": Array [],
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "name": "New Item 2",
        "price": 2,
        "tags": Array [],
      },
    ],
  },
  Object {
    "items": Array [],
  },
]
`;

exports[`can rename an item in the cart 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "name": "New Item 1",
      "price": 1,
      "tags": Array [],
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "New Item (new)",
  },
]
`;

exports[`should create a cart instance 1`] = `
Object {
  "items": Array [
    Object {
      "name": "Instance Item",
      "price": 2.83,
      "tags": Array [
        Object {
          "name": "first",
        },
      ],
    },
  ],
}
`;

exports[`should create a model instance 1`] = `
Array [
  Object {
    "op": "replace",
    "path": "/name",
    "value": "Google Pixel 4",
  },
  Object {
    "op": "replace",
    "path": "/price",
    "value": 799,
  },
]
`;
